File: python/pages/demoqa_home_page.py
from playwright.sync_api import Page, expect
from helpers.book_data import first_run_values
from datetime import datetime

class DemoqaHomePage:

    def __init__(self, page: Page):
        self.page = page
        self.logo_home = page.locator('div#app header a')
        self.input_box = page.locator('input#searchBox')
        self.result = page.locator('span a')
        self.drop_down = page.get_by_label('rows per page')
        self.button_previous = page.get_by_role("button", name="Previous")
        self.button_next = page.get_by_role("button", name="Next")
    
    def go_to(self, url):
        self.page.goto(url)
    
    def check_home_page(self):
        self.logo_home.is_visible()
        self.input_box.is_visible()

    def enter_book(self, book_name):
        self.input_box.fill(book_name)

    def check_rows_amount(self):
        row_count = self.result
        count = row_count.count()
        expected_row = self.result.all()
        print(f"this is count : {count}");
        assert len(expected_row) == 1
        # self.result.count()
        expect(self.result).to_have_text("Git Pocket Guide")

    def get_book_text(self, run: str):
        results = self.result
        # expect(self.result).to_be_visible()
        count = results.count()

        # expected_values = first_run_values if run == "First" else second_run_values

        for i in range(count):
            text = results.nth(i).text_content()
            # assert text == expected_values[i], f"item {i} mismatch: 'expected_values[i]', got '{text}'"
            assert text == first_run_values[i]
            print(text)

            date_str = datetime.now().strftime("%Y-%m-%d")
            content = self.result.all_inner_texts()
            filename = f"output_{date_str}.txt"
    
    # Write to the dynamic file
            with open(filename, "w", encoding="utf-8") as file:
                file.write(f"{content}\n")

            print(f"Items written to {filename}")

File: playwright-ts-FE/pages/demoqaHome.page.ts
import { expect, Locator, Page } from '@playwright/test';
import { firstRunValues, secondRunValues } from '../helpers/constants';
import { clickIfElementClickable, isButtonClickable } from '../helpers/helpers';
import fs from 'fs';

export class DemoqaHomePage {
    readonly page: Page;
    readonly logoHome: Locator;
    readonly inputBox: Locator;
    readonly result: Locator;
    readonly dropdown: Locator;
    readonly selectValue: Locator;
    readonly buttonPrevious: Locator;
    readonly buttonNext: Locator;

    constructor(page: Page) {
        this.page = page;
        this.logoHome = this.page.locator('div#app header a');
        this.inputBox = this.page.locator('input#searchBox');
        this.result = this.page.locator('span a');
        this.dropdown = this.page.getByLabel('rows per page');
        this.selectValue = this.page.locator('span select option');
        this.buttonPrevious = this.page.getByRole('button', { name: 'Previous' });
        this.buttonNext = this.page.getByRole('button', { name: 'Next' });
    }

    async goToHome(url: string): Promise<void> {
        await this.page.goto(url);
    }

    async checkHomePage(): Promise<void> {
        await expect(this.logoHome).toBeVisible();
        await expect(this.inputBox).toBeVisible();

    }

    async searchForBoook(title: string): Promise<void> {
        await this.inputBox.fill(title);
    }

    async checkResult(title: string): Promise<void> {
        await expect(this.result).toBeVisible();
        await expect(this.result).toHaveText(title);
    }

    async checkNumberOfRows(): Promise<void> {
        const numberOfRows = await this.result.all();
        expect(numberOfRows.length).toEqual(1);
    }

    async goToBookDetail(nthElement: number): Promise<void> {
        await expect(this.result.nth(nthElement)).toBeVisible();
        await this.result.nth(nthElement).click();
    }

    async checkUrl(expectedUrl: string): Promise<void> {
        await expect(this.page).toHaveURL(expectedUrl);
    }

    async changeNumberOfRows(): Promise<void> {
        await expect(this.dropdown).toBeVisible();
        await this.dropdown.click();
        await this.dropdown.selectOption('5');
        // await this.page.waitForTimeout(1000);
    }

    async clickIfEnabled(locator: 'Next'| 'Previous'): Promise<void> {
        (locator === 'Next') ? await clickIfElementClickable(this.buttonNext) : await clickIfElementClickable(this.buttonPrevious) 
    }

    async clickIfEnabledTwo(): Promise<void> {
        await clickIfElementClickable(this.buttonPrevious)
    }

        async clickIfEnabledThree(): Promise<void> {
        (await isButtonClickable(this.buttonPrevious) === true) ? await this.buttonPrevious.click() : await this.buttonNext.click();
    }

    async clickNextButton(): Promise<any> {
        // await this.page.waitForTimeout(2000);
        const bookTitle = await this.result.first().textContent();
        console.info(`Book title: ${bookTitle}`);
        fs.writeFileSync('extractedBookTitle.txt', bookTitle || 'No title found', 'utf-8');
        return bookTitle;
    }

    async clickPreviousButton(titleOne: string): Promise<void> {
        // await this.page.waitForTimeout(1000);
        await expect(this.result.first()).toBeVisible();
        const bookTitle = await this.result.first().textContent();
        expect(bookTitle).not.toEqual(titleOne);
    }

    async getBookList(run: 'First' | 'Second'): Promise<void> {
        await expect(this.result.nth(2)).toBeVisible();

        const amount = await this.result.count();

        for (let i = 0; i < amount ; i++) {
            // console.info(await this.result.nth(i).textContent());

            (run === 'First') ? expect(await this.result.nth(i).textContent()).toEqual(firstRunValues[i]) : expect(await this.result.nth(i).textContent()).toEqual(secondRunValues[i]);
        }
    }
}